core/src/main/java/io/grpc/ClientInterceptor.java

```diff
@@ -31,35 +31,74 @@
 
 package io.grpc;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
- * Interface for intercepting outgoing calls before they are dispatched by a {@link Channel}.
+ * A {@link CallFactory} for intercepting outgoing calls before they are dispatched by a {@link
+ * CallFactory}.
  *
- * <p>Implementers use this mechanism to add cross-cutting behavior to {@link Channel} and
- * stub implementations. Common examples of such behavior include:
- * <ul>
- * <li>Adding credentials to header metadata</li>
- * <li>Logging and monitoring call behavior</li>
- * <li>Request and response rewriting</li>
- * </ul>
+ * <p>Implementers use this mechanism to add cross-cutting behavior to {@link CallFactory} and stub
+ * implementations. Common examples of such behavior include: <ul> <li>Adding credentials to header
+ * metadata</li> <li>Logging and monitoring call behavior</li> <li>Request and response
+ * rewriting</li> </ul>
  */
 @ThreadSafe
-public interface ClientInterceptor {
+public abstract class ClientInterceptor extends CallFactory {
+  private CallFactory next;
```

Comment:
Agreed that this implementation is a bit hacky :).

Actually the more I think about it, I don't think what we really need is an `InterceptorFactory`, not `Interceptor`. The call to create an interceptor chain should just take a list of these factories which will look like:

```java
interface InterceptorFactory {
    ClientCallFactory newInterceptor(ClientCallFactory next);
}
```

I'll sketch this out and update the PR shortly.